src/datahike/db.cljc

```diff
@@ -1375,277 +1495,339 @@
+                        (if (= v ov)
+                          (recur (transact-add report [:db/add e a nv]) entities)
+                          (raise ":db.fn/cas failed on datom [" e " " a " " v "], expected " ov
+                            {:error :transact/cas, :old (first datoms), :expected ov, :new nv})))))))
+
+              (tx-id? e)
+              (recur (allocate-eid report e (current-tx report)) (cons [op (current-tx report) a v] entities))
+
+              (and (ref? db a) (tx-id? v))
+              (recur (allocate-eid report v (current-tx report)) (cons [op e a (current-tx report)] entities))
+
+              (tempid? e)
+              (let [upserted-eid (when (is-attr? db a :db.unique/identity)
+                                   (:e (first (-datoms db :avet [a v]))))
+                    allocated-eid (get tempids e)]
+                (if (and upserted-eid allocated-eid (not= upserted-eid allocated-eid))
+                  (retry-with-tempid initial-report report initial-es e upserted-eid)
+                  (let [eid (or upserted-eid allocated-eid (next-eid db))]
+                    (recur (allocate-eid report e eid) (cons [op eid a v] entities)))))
+
+              (and (ref? db a) (tempid? v))
+              (if-let [vid (get tempids v)]
+                (recur report (cons [op e a vid] entities))
+                (recur (allocate-eid report v (next-eid db)) es))
+
+              (and (homogeneous-tuple-attr? db a)
+                (> (count v) 8))
+              (raise "Cannot store more than 8 values for homogeneous tuple: " entity
+                {:error :transact/syntax, :tx-data entity})
+
+              (and (homogeneous-tuple-attr? db a)
+                (not (apply = (map type v))))
+              (raise "Cannot store homogeneous tuple with values of different type: " entity
+                {:error :transact/syntax, :tx-data entity})
+
+              (and (homogeneous-tuple-attr? db a)
+                (not (s/valid? (-> db -schema a :db/tupleType) (first v))))
+              (raise "Cannot store homogeneous tuple. Values are of wrong type: " entity
+                {:error :transact/syntax, :tx-data entity})
+
+              (and (heterogeneous-tuple-attr? db a)
+                (not (= (count v) (count (-> db -schema a :db/tupleTypes)))))
+              (raise (str "Cannot store heterogeneous tuple: expecting " (count (-> db -schema a :db/tupleTypes)) " values, got " (count v))
+                {:error :transact/syntax, :tx-data entity})
+
+              (and (heterogeneous-tuple-attr? db a)
+                (not (apply = (map s/valid? (-> db -schema a :db/tupleTypes) v))))
+              (raise (str "Cannot store heterogeneous tuple: there is a mismatch between values " v " and their types " (-> db -schema a :db/tupleTypes))
+                {:error :transact/syntax, :tx-data entity})
+
+              (and (not (::internal (meta entity)))
+                (composite-tuple-attr? db a))
```

Comment:
I'd love to see this the other way round just for the sake of consistency with the above blocks.

Reply:
```
(and (composite-tuple...
    (not (::internal...
```